tests/integration/targets/test_proxysql_query_rules_fast_routing/tasks/main.yml

```
@@ -19,11 +19,18 @@
 - name: "{{ role_name }} | test_create_query_rule_fast_routing | test create query rule fast routing"
   import_tasks: test_create_query_rule_fast_routing.yml
   vars:
+    test_destination_hostgroup: 1
     test_proxysql_query_rules_fast_routing_cleanup_after_test: false
 - name: "{{ role_name }} | test_create_query_rule_fast_routing | test idempotence of create query rule fast routing"
   import_tasks: test_create_query_rule_fast_routing.yml
   vars:
+    test_destination_hostgroup: 1
+    test_proxysql_query_rules_fast_routing_cleanup_after_test: false
     test_proxysql_query_rules_fast_routing_check_idempotence: true
+- name: "{{ role_name }} | test_create_query_rule_fast_routing | test update query rule fast routing"
+  import_tasks: test_create_query_rule_fast_routing.yml
+  vars:
+    test_destination_hostgroup: 2
 
```

```suggestion
    test_destination_hostgroup: 2
    test_proxysql_query_rules_fast_routing_check_idempotence: true

```

otherwise the `base_test.yml` still clean up the table ... it's very hard to debug :D

---

This case expects status `changed`, so it should be `test_proxysql_query_rules_fast_routing_check_idempotence: false`.

---

`idempotence` in this case means, that the task with the parameters are applied twice, and the previous values will not be deleted/purged.  
When it is set to `false`, it will purge the table: https://github.com/ansible-collections/community.proxysql/blob/main/tests/integration/targets/test_proxysql_query_rules_fast_routing/tasks/base_test.yml#L8-L9

You can verify it. Just apply the integration test to the main branch, without the fix. The integration test will not fail - but it must fail.

```
TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "msg": "######################## BEFORE"
}

TASK [test_proxysql_query_rules_fast_routing : Create fast routing rules] ******
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "out.mysql_query_rules_fast_routing": [
        {
            "comment": "testing fast route",
            "destination_hostgroup": "1",
            "flagIN": "0",
            "schemaname": "default",
            "username": "user_ro"
        }
    ]
}

TASK [test_proxysql_query_rules_fast_routing : include_tasks] ******************
included: /root/ansible_collections/community/proxysql/tests/output/.tmp/integration/test_proxysql_query_rules_fast_routing-g0ys72kn-ÅÑŚÌβŁÈ/tests/integration/targets/test_proxysql_query_rules_fast_routing/tasks/base_test.yml for testhost

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | are we performing a delete] ***
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure we're in a clean state when we start] ***
included: /root/ansible_collections/community/proxysql/tests/output/.tmp/integration/test_proxysql_query_rules_fast_routing-g0ys72kn-ÅÑŚÌβŁÈ/tests/integration/targets/test_proxysql_query_rules_fast_routing/tasks/cleanup_test_query_rules_fast_routing.yml for testhost

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure no query rules fast routing are created] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure no query rules fast routing are saved on disk] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure no query rules fast routing are loaded to runtime] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "msg": "########################## IN BASIC TEST"
}

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "test_delete": false
}

TASK [test_proxysql_query_rules_fast_routing : Create fast routing rules] ******
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "out.mysql_query_rules_fast_routing": []
}

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | create test query rules fast routing] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | save the query rules fast routing config from memory to disk] ***
skipping: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | load the query rules fast routing config from memory to runtime] ***
skipping: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | check if test query rule fast routing exists in memory] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | check if test query rule fast routing exists on disk] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | check if test query rule fast routing exists in runtime] ***
changed: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | check if create query rule fast routing reported a change] ***
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | confirm create query rule fast routing did make a change in memory] ***
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | confirm create query rule fast routing did make a change on disk] ***
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | confirm create query rule fast routing did make a change to runtime] ***
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "msg": "######################### DONE"
}

TASK [test_proxysql_query_rules_fast_routing : Create fast routing rules] ******
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "out.mysql_query_rules_fast_routing": [
        {
            "comment": "testing fast route",
            "destination_hostgroup": "2",
            "flagIN": "0",
            "schemaname": "default",
            "username": "user_ro"
        }
    ]
}

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure no query rules fast routing are created] ***
skipping: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure no query rules fast routing are saved on disk] ***
skipping: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure no query rules fast routing are loaded to runtime] ***
skipping: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "msg": "################# after cleanup"
}

TASK [test_proxysql_query_rules_fast_routing : Create fast routing rules] ******
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "out.mysql_query_rules_fast_routing": [
        {
            "comment": "testing fast route",
            "destination_hostgroup": "2",
            "flagIN": "0",
            "schemaname": "default",
            "username": "user_ro"
        }
    ]
}

```

When you add `    test_proxysql_query_rules_fast_routing_check_idempotence: true`, it fail in the main branch, but not in your branch/PR.

```
TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "msg": "######################## BEFORE"
}

TASK [test_proxysql_query_rules_fast_routing : Create fast routing rules] ******
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "out.mysql_query_rules_fast_routing": [
        {
            "comment": "testing fast route",
            "destination_hostgroup": "1",
            "flagIN": "0",
            "schemaname": "default",
            "username": "user_ro"
        }
    ]
}

TASK [test_proxysql_query_rules_fast_routing : include_tasks] ******************
included: /root/ansible_collections/community/proxysql/tests/output/.tmp/integration/test_proxysql_query_rules_fast_routing-wiren2tq-ÅÑŚÌβŁÈ/tests/integration/targets/test_proxysql_query_rules_fast_routing/tasks/base_test.yml for testhost

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | are we performing a delete] ***
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | ensure we're in a clean state when we start] ***
skipping: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "msg": "########################## IN BASIC TEST"
}

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "test_delete": false
}

TASK [test_proxysql_query_rules_fast_routing : Create fast routing rules] ******
ok: [testhost]

TASK [test_proxysql_query_rules_fast_routing : debug] **************************
ok: [testhost] => {
    "out.mysql_query_rules_fast_routing": [
        {
            "comment": "testing fast route",
            "destination_hostgroup": "1",
            "flagIN": "0",
            "schemaname": "default",
            "username": "user_ro"
        }
    ]
}

TASK [test_proxysql_query_rules_fast_routing : test_proxysql_query_rules_fast_routing | test_create_query_rule_fast_routing | create test query rules fast routing] ***
fatal: [testhost]: FAILED! => {"changed": false, "msg": "unable to modify rule: (1045, 'ProxySQL Admin Error: UNIQUE constraint failed: mysql_query_rules_fast_routing.username, mysql_query_rules_fast_routing.schemaname, mysql_query_rules_fast_routing.flagIN')"}
```

---

here is .patch file you can apply on your main branch (_not your PR branch_). https://gist.github.com/markuman/1614e3283651ad0d2dd681d6af6a9f23

---

it reduces the test to the relevant part.

`ansible-test integration test_proxysql_query_rules_fast_routing --docker default`

---

https://github.com/ansible-collections/community.proxysql/blob/main/tests/integration/targets/test_proxysql_query_rules_fast_routing/tasks/base_test.yml#L8-L9
Here asserts that the result is `not changed` when `test_proxysql_query_rules_fast_routing_check_idempotence` is  `true`, but it should report `changed`.

Maybe it time to separate the expected result and

---

I would chose to set a `test_proxysql_query_rules_fast_routing_cleanup_before_test` flag for triggering the cleanup task and left `idempotence` to do assertion stuff.

---

@squirrel532  That works for me too.

---

The `destination_hostgroup` still remains  `1` in assert task, and I not sure what happened. i'm still working on it.

---

> and I not sure what happened. i'm still working on it.

that's why I suggest to place a new task file that covers only that case @squirrel532 :)